File: app/src/main/java/com/example/bikeridedetection/service/BikeTransitionReceiver.java
package com.example.bikeridedetection.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;

import com.example.bikeridedetection.domain.repository.BikeModeRepository;
import com.google.android.gms.location.ActivityTransition;
import com.google.android.gms.location.ActivityTransitionEvent;
import com.google.android.gms.location.ActivityTransitionResult;

import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.Dispatchers;
import kotlinx.coroutines.Job;
import kotlinx.coroutines.SupervisorKt;
import timber.log.Timber;

/**
 * Broadcast receiver for handling bike activity transitions.
 * Automatically enables/disables bike mode based on detected activity.
 */
@AndroidEntryPoint
public class BikeTransitionReceiver extends BroadcastReceiver {

    public static final String ACTION_BIKE_MODE_CHANGED = "com.example.bikeridedetection.BIKE_MODE_CHANGED";
    public static final String KEY_BIKE_MODE = "bike_mode";

    @Inject
    BikeModeRepository bikeModeRepository;

    private final Job supervisorJob = SupervisorKt.SupervisorJob(null);
    private final CoroutineScope receiverScope = CoroutineScopeHelper.createScope(supervisorJob, Dispatchers.getIO());

    @Override
    public void onReceive(@NonNull Context context, @NonNull Intent intent) {
        if (!ActivityTransitionResult.hasResult(intent)) {
            return;
        }

        ActivityTransitionResult result = ActivityTransitionResult.extractResult(intent);
        if (result == null) {
            return;
        }

        PendingResult pendingResult = goAsync();
        List<ActivityTransitionEvent> events = result.getTransitionEvents();

        if (events.isEmpty()) {
            pendingResult.finish();
            return;
        }

        AtomicInteger pendingCount = new AtomicInteger(events.size());

        for (ActivityTransitionEvent event : events) {
            Timber.d("Detected transition: type=%d, transition=%d",
                    event.getActivityType(), event.getTransitionType());

            boolean bikeModeOn = event.getTransitionType() == ActivityTransition.ACTIVITY_TRANSITION_ENTER;

            CoroutineScopeHelper.handleTransitionAsync(
                    receiverScope,
                    bikeModeRepository,
                    bikeModeOn,
                    () -> {
                        Timber.d("Bike mode toggled: %s", bikeModeOn);

                        // Broadcast to other components
                        Intent broadcast = new Intent(ACTION_BIKE_MODE_CHANGED);
                        broadcast.putExtra(KEY_BIKE_MODE, bikeModeOn);
                        broadcast.setPackage(context.getPackageName());
                        context.sendBroadcast(broadcast);

                        // Finish when all events are processed
                        if (pendingCount.decrementAndGet() == 0) {
                            pendingResult.finish();
                        }
                    }
            );
        }
    }
}

File: app/src/main/java/com/example/bikeridedetection/service/BikeTransitionReceiver.kt
package com.example.bikeridedetection.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.example.bikeridedetection.domain.repository.BikeModeRepository
import com.google.android.gms.location.ActivityTransition
import com.google.android.gms.location.ActivityTransitionResult
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * Broadcast receiver for handling bike activity transitions.
 * Automatically enables/disables bike mode based on detected activity.
 */
@AndroidEntryPoint
class BikeTransitionReceiver : BroadcastReceiver() {
    @Inject
    lateinit var bikeModeRepository: BikeModeRepository

    private val receiverScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    override fun onReceive(
        context: Context,
        intent: Intent,
    ) {
        if (!ActivityTransitionResult.hasResult(intent)) {
            return
        }

        val result = ActivityTransitionResult.extractResult(intent) ?: return
        val pendingResult = goAsync()

        receiverScope.launch {
            try {
                result.transitionEvents.forEach { event ->
                    Timber.d(
                        "Detected transition: type=${event.activityType}, " +
                            "transition=${event.transitionType}",
                    )

                    val bikeModeOn = event.transitionType == ActivityTransition.ACTIVITY_TRANSITION_ENTER
                    bikeModeRepository.setBikeModeEnabled(bikeModeOn)

                    Timber.d("Bike mode toggled: $bikeModeOn")

                    // Broadcast to other components
                    val broadcast =
                        Intent(ACTION_BIKE_MODE_CHANGED).apply {
                            putExtra(KEY_BIKE_MODE, bikeModeOn)
                            setPackage(context.packageName)
                        }
                    context.sendBroadcast(broadcast)
                }
            } finally {
                pendingResult.finish()
            }
        }
    }

    companion object {
        const val ACTION_BIKE_MODE_CHANGED = "com.example.bikeridedetection.BIKE_MODE_CHANGED"
        const val KEY_BIKE_MODE = "bike_mode"
    }
}
